### PR TITLE
New models

### PR DESCRIPTION
**Pull request by gbalduzz**
- Multiband models used by me and Peizhi.
- Changed the symmetrization code to decide which symmetries should be considered by the model, with the option of declaring an antisymmetry.
- Fixed bug in Model::get_flavours() initialization.

### CODE REVIEW

**Comment by gbalduzz** [comment]
retest this, please.

File: include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+}
+
+template <typename PointGroup>
+double* TwobandCu<PointGroup>::initialize_k_LDA_basis() {
+  static std::array<double, 4> basis{2 * M_PI, 0, 0, 2 * M_PI};
+  return basis.data();
+}
+
+template <typename PointGroup>
+std::vector<int> TwobandCu<PointGroup>::get_flavors() {
```
Comment (PDoakORNL) [inline include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp:139]: none of these should be get_ they are not simple getters or setters in fact they wrap singleton like vectors.

File: include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+  k_LDA[0] = 2. * M_PI;
+  k_LDA[1] = 0.;
+  k_LDA[2] = 0.;
+  k_LDA[3] = 2. * M_PI;
+
+  return k_LDA;
+}
+
+template <typename point_group_type>
+std::vector<int> threeband_hubbard<point_group_type>::get_flavors() {
```
Comment (PDoakORNL) [inline include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp:171]: this a weird pattern. should not be get_ as it is not a simple getter or setter flavors is in fact not even a member variable.  Why is it important to have just one of this very small object? Why should it be global?  It looks like it should be a member variable. Are we just following a questionable convention an earlier developer used?

```C++
const std::vector<int> flavors{0,1,1};
```

perhaps static if you insist.

File: include/dca/phys/models/traits.hpp [hunk 3]
```diff
@@ -38,10 +38,10 @@ struct has_non_density_interaction<FeAsLattice<BaseLattice>> {
   constexpr static bool value = true;
 };
 
-//template <class PointGroup>
-//struct has_non_density_interaction<TwobandCu<PointGroup>> {
-//  constexpr static bool value = true;
-//};
+template <class PointGroup>
+struct has_non_density_interaction<TwobandCu<PointGroup>> {
```
Comment (PDoakORNL) [inline include/dca/phys/models/traits.hpp:42]: This is a special naming case we haven't considered for now I think we should leave it alone. It is really a type but its either a query function or a "variable" as a metaprogramming entity.

File: include/dca/phys/parameters/model_parameters_threeband_hubbard.inc [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+  int getBufferSize(const Concurrency& concurrency) const;
+  template <typename Concurrency>
+  void pack(const Concurrency& concurrency, char* buffer, int buffer_size, int& position) const;
+  template <typename Concurrency>
+  void unpack(const Concurrency& concurrency, char* buffer, int buffer_size, int& position);
+
+  template <typename ReaderOrWriter>
+  void readWrite(ReaderOrWriter& reader_or_writer);
+
+  double get_t_pd() const {
```
Comment (PDoakORNL) [inline include/dca/phys/parameters/model_parameters_threeband_hubbard.inc:28]: This is an acceptable get_ method.

File: include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+  const static ClusterShapeType DCA_cluster_shape = BETT_CLUSTER;
+  const static ClusterShapeType LDA_cluster_shape = PARALLELEPIPED;
+
+  const static int DIMENSION = 2;
+  const static int BANDS = 2;
+
+  // Rotations of pi/2 are an anti-symmetry on the band off-diagonal.
+  static int transformationSign(int b1, int b2, int s);
+
+  static double* initialize_r_DCA_basis();
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp:49]: Unfortunately, this method is part of the interface of a model. I will see if I can easily refactor everything with sed.

File: include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp (hunk 2 above)
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp:171]: this is again part of the old model interface. Will see if we can do better.

File: include/dca/phys/models/traits.hpp (hunk 3 above)
Comment (gbalduzz) [reply include/dca/phys/models/traits.hpp:42]: actually we can make it a real variable with c++17. Will update.

File: include/dca/phys/dca_step/cluster_solver/exact_diagonalization_advanced/hamiltonian.hpp [hunk 6]
```diff
@@ -234,7 +236,8 @@ void Hamiltonian<parameters_type, ed_options>::initialize(
                              func::dmn_variadic<func::dmn_variadic<b_dmn, s_dmn>,
                                                 func::dmn_variadic<b_dmn, s_dmn>, RClusterDmn>>& H_0,
     /*const*/ func::function<double, func::dmn_variadic<func::dmn_variadic<b_dmn, s_dmn>,
-                                                        func::dmn_variadic<b_dmn, s_dmn>, RClusterDmn>>& H_int) {
+                                                        func::dmn_variadic<b_dmn, s_dmn>, RClusterDmn>>&
+        H_int) {
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/exact_diagonalization_advanced/hamiltonian.hpp:240]: yes, I run it. This is how it chose to split the line.

File: include/dca/phys/dca_step/cluster_solver/exact_diagonalization_advanced/greens_functions/sp_greens_function.hpp [hunk 7]
```diff
@@ -107,19 +107,19 @@ class SpGreensFunction {
...
   void renormalize_real_space_Greens_functions(
       func::function<std::complex<double>, func::dmn_variadic<nu_dmn, nu_dmn, RClusterDmn, w>>& G_r_w_im,
       func::function<std::complex<double>, func::dmn_variadic<nu_dmn, nu_dmn, RClusterDmn, w_REAL>>& G_r_w_re,
       func::function<double, func::dmn_variadic<nu_dmn, nu_dmn, RClusterDmn, t>>& G_r_t,
-      func::function<complex_type, func::dmn_variadic<nu_dmn, nu_dmn, RClusterDmn, RClusterDmn, w_VERTEX, w_VERTEX>>&
-          G_nu_nu_r_r_w_w,
-      func::function<complex_type, func::dmn_variadic<nu_dmn, nu_dmn, KClusterDmn, KClusterDmn, w_VERTEX, w_VERTEX>>&
-          G_nu_nu_k_k_w_w);
+      func::function<complex_type, func::dmn_variadic<nu_dmn, nu_dmn, RClusterDmn, RClusterDmn,
+                                                      w_VERTEX, w_VERTEX>>& G_nu_nu_r_r_w_w,
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/exact_diagonalization_advanced/greens_functions/sp_greens_function.hpp:120]: cleaned up in 6bf1d40ac7713157d77c107652904c321ff4277b

File: include/dca/phys/dca_step/lattice_mapping/lattice_mapping_sp.hpp [hunk 8]
```diff
@@ -70,8 +71,8 @@ class lattice_mapping_sp {
   void plot_function(func::function<std::complex<double>, func::dmn_variadic<nu, nu, k_dmn_t, w>>& f);
 
 private:
-  parameters_type& parameters;
-  concurrency_type& concurrency;
+  /*const*/ parameters_type& parameters;
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/lattice_mapping/lattice_mapping_sp.hpp:74]: This should truly be const. Unfortunately, all the objects down the line have no const flags.

File: include/dca/phys/dca_data/dca_data.hpp [hunk 9]
```diff
@@ -122,8 +122,8 @@ class DcaData {
   void print_Sigma_QMC_versus_Sigma_cg();
 
 private:
-  Parameters& parameters_;
-  Concurrency& concurrency_;
+  /*const*/ Parameters& parameters_;
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_data/dca_data.hpp]: in 571 the chemical potential of the parameters is updated, which is a rather unfortunate choice as this is the only non-const member of the Parameter class. But I will try to fix this later.
Removed /*const*/ in 2da80c2782bd90dc17bcd5604147d23d8aa3fd09

File: include/dca/phys/dca_step/symmetrization/symmetrize_single_particle_function.hpp [hunk 10]
```diff
@@ -466,49 +471,49 @@ void symmetrize_single_particle_function::executeTimeOrFreq(
...
+    func::function<scalartype, func::dmn_variadic<BDmn, BDmn, WRealDmn>>& /*f*/, bool /*do_diff*/) {}
 
 template <typename scalartype>
-void symmetrize_single_particle_function::execute(func::function<scalartype, w_VERTEX>& f,
+void symmetrize_single_particle_function::execute(func::function<scalartype, WVertexDmn>& f,
                                                   bool do_diff) {
   double max = 0;
-  for (int i = 0; i < w_VERTEX::dmn_size() / 2; i++) {
-    max = std::max(max, abs((f(i) - std::conj(f(w_VERTEX::dmn_size() - i - 1))) / 2.));
+  for (int i = 0; i < WVertexDmn::dmn_size() / 2; i++) {
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/dca_step/symmetrization/symmetrize_single_particle_function.hpp:489]: run sed in 2da80c2782bd90dc17bcd5604147d23d8aa3fd09.

File: include/dca/phys/models/analytic_hamiltonians/singleband_chain.hpp [hunk 11]
```diff
@@ -42,6 +42,13 @@ class singleband_chain {
   const static int DIMENSION = 2;
   const static int BANDS = 1;
 
+  constexpr static int transformationSignofR(int, int, int) {
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/singleband_chain.hpp]: No, this is a tool to select symmetry/antisymmetry/ignore. #184 is about the Monte Carlo sign problem.

File: include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+#include "dca/util/type_list.hpp"
+
+namespace dca {
+namespace phys {
+namespace models {
+// dca::phys::models::
+
+// TODO: the symmetry of this system must be checked.
+template <typename symmetry_group>
+class threeband_hubbard {
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp]: fixed in ad2485498f0d2e09b41105fb887dedbeb9e90280

File: include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+
+  static std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>> get_orbital_permutations();
+
+  // Rotations of pi/2 are an anti-symmetry on the band off-diagonal.
+  static int transformationSignofR(int b1, int b2, int s);
+  static int transformationSignofK(int b1, int b2, int s);
+
+  // Initializes the interaction Hamiltonian in real space.
+  template <typename BandDmn, typename SpinDmn, typename RDmn, typename parameters_type>
+  static void initialize_H_interaction(
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp]: update in ad2485498f0d2e09b41105fb887dedbeb9e90280

File: include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+
+          if (i == j && i != 0 && s1 != s2)
+            H_interaction(i, s1, j, s2, origin) = U_pp;
+        }
+      }
+    }
+  }
+}
+
+template <typename point_group_type>
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp]: fixed in ad2485498f0d2e09b41105fb887dedbeb9e90280

File: include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+#include "dca/phys/models/analytic_hamiltonians/cluster_shape_type.hpp"
+#include "dca/util/type_list.hpp"
+
+namespace dca {
+namespace phys {
+namespace models {
+// dca::phys::models::
+
+template <typename PointGroup>
+class TwobandCu {
```
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp]: fixed it in ad2485498f0d2e09b41105fb887dedbeb9e90280

File: include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp]: run sed in ad2485498f0d2e09b41105fb887dedbeb9e90280

File: include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp (hunk 1 above)
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/twoband_Cu.hpp]: updated the names in ad2485498f0d2e09b41105fb887dedbeb9e90280

File: include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp (hunk 2 above)
Comment (gbalduzz) [reply include/dca/phys/models/analytic_hamiltonians/threeband_hubbard.hpp]: changed to brace list initialization in ad2485498f0d2e09b41105fb887dedbeb9e90280

File: include/dca/phys/models/traits.hpp (hunk 3 above)
Comment (gbalduzz) [reply include/dca/phys/models/traits.hpp]: done in bc0b2c91a9581e3591c52157db8bd1de90312ae9

**Comment by PDoakORNL** [comment]
Doing a final test build

**Pull request merged**